proposals/poc/witx/proposal_symmetric.md

```diff
@@ -0,0 +1,1418 @@
+The requested feature is not supported by the chosen algorithm.
+
+- <a href="#crypto_errno.prohibited_operation" name="crypto_errno.prohibited_operation"></a> `prohibited_operation`
+The requested operation is valid, but was administratively prohibited.
+
+- <a href="#crypto_errno.unsupported_encoding" name="crypto_errno.unsupported_encoding"></a> `unsupported_encoding`
+Unsupported encoding for an import or export operation.
+
+- <a href="#crypto_errno.unsupported_algorithm" name="crypto_errno.unsupported_algorithm"></a> `unsupported_algorithm`
+The requested algorithm is not supported by the host.
+
+- <a href="#crypto_errno.unsupported_option" name="crypto_errno.unsupported_option"></a> `unsupported_option`
+The requested option is not supported by the currently selected algorithm.
+
+- <a href="#crypto_errno.invalid_key" name="crypto_errno.invalid_key"></a> `invalid_key`
+An invalid or incompatible key was supplied.
+
+The key may not be valid, or was generated for a different algorithm or parameters set.
+
+- <a href="#crypto_errno.invalid_length" name="crypto_errno.invalid_length"></a> `invalid_length`
+The currently selected algorithm doesn't support the requested output length.
+
+This error is thrown by non-extensible hash functions, when requesting an output size larger than they produce out of a single block.
+
+- <a href="#crypto_errno.verification_failed" name="crypto_errno.verification_failed"></a> `verification_failed`
+A signature or authentication tag verification failed.
+
+- <a href="#crypto_errno.rng_error" name="crypto_errno.rng_error"></a> `rng_error`
+A secure random numbers generator is not available.
+
+The requested operation requires random numbers, but the host cannot securely generate them at the moment.
+
+- <a href="#crypto_errno.algorithm_failure" name="crypto_errno.algorithm_failure"></a> `algorithm_failure`
+An error was returned by the underlying cryptography library.
+
+The host may be running out of memory, parameters may be incompatible with the chosen implementation of an algorithm or another unexpected error may have happened.
+
+Ideally, the specification should provide enough details and guidance to make this error impossible to ever be thrown.
+
+Realistically, the WASI crypto module cannot possibly cover all possible error types implementations can return, especially since some of these may be language-specific.
+This error can thus be thrown when other error types are not suitable, and when the original error comes from the cryptographic primitives themselves and not from the WASI module.
+
+- <a href="#crypto_errno.invalid_signature" name="crypto_errno.invalid_signature"></a> `invalid_signature`
+The supplied signature is invalid, or incompatible with the chosen algorithm.
+
+- <a href="#crypto_errno.closed" name="crypto_errno.closed"></a> `closed`
+An attempt was made to close a handle that was already closed.
```

Comment:
This seems like a slippery slope, because -- I assume -- handle values might also be reused.

Reply:
Actually not!

In POSIX interfaces, descriptors are reused, only to remain compatible with the legacy `select(2)` system call.

WASI handles don't have that constraint. And in the current implementation, it is actually quite hard to get them reused even by accident. Closing a handle removes it from the table, but identifiers keep being incremented. The same identifier is not going to be reused before a full cycle. Also, the top bits of a handle encode the object type. This is extremely useful for debugging (when `invalid handle` is returned, one can immediately check if this is a type issue), but also prevents the same handle identifier to ever represent different types, even on reuse.